### PR TITLE
add pre validate step (check snapshot names)

### LINKED ISSUE #14: hcloud builder creates multiple image snapshots with the same name
This is the same problem that the Scaleway builder had (https://github.com/hashicorp/packer/issues/9839), for which I wrote a PR (https://github.com/hashicorp/packer/pull/9840) that has been merged. In the same spirit, I would like to propose a PR also in this case.

#### Overview of the Issue

If you set the hcloud builder parameter [snapshot_name](https://www.packer.io/docs/builders/hetzner-cloud#snapshot_name) to a fixed string, and you run Packer more than once, it will create multiple snapshots with the same name. 
Note: it never creates "images", or, said in another way, user-created "images" (what then you would pass to Terraform) are always "snapshots".

Although these multiple snapshots are distinguished by the backend with a unique ID, for example:

```
$ hcloud image list | grep snapshot
ID       TYPE     NAME DESCRIPTION  ...  CREATED
49206576 snapshot -    banana       ...  Wed Sep 29 20:07:02 CEST 2021
49206827 snapshot -    banana       ...  Wed Sep 29 20:08:30 CEST 2021
```

from a Packer point of view, in my understanding this should not happen. I am used to the Packer AWS EC2 builder and rightly so it refuses to build an image if another one exists with the same name, see builder/amazon/common/step_pre_validate.go.

From my understanding, the current behavior (multiple images/snapshots with the same name) should happen only if the `-force` flag is specified to `packer build`.

Consider also that Hetzener Cloud makes you pay each snapshot based on the image size, and normally in Terraform you would refer to it with the "name" (so getting always the latest in the series), not with the unique ID.

From my point of view, this behavior is both error-prone and a waste of money: each time you run a `packer build`, you _add_ to the bill :-)

#### Reproduction Steps

Run `packer build` at least twice:

```
$ packer build repro.pkr.hcl
$ packer build repro.pkr.hcl
```

Verify in the hetzner cloud console (or with the command-line tool `hcloud image list` that there are two snapshots with the same name.

### Plugin and Packer version

Packer v1.7.4
hcloud plugin latest version v1.0.1

### Simplified Packer Buildfile

```
packer {
  required_version = ">= 1.7.4"
  required_plugins {
    hcloud = {
      version = ">= 1.0.1"
      source  = "github.com/hashicorp/hcloud"
    }
  }
}

source "hcloud" "cfg" {
  # set token via env var HCLOUD_TOKEN
  image        = "debian-11"
  location     = "hel1"
  server_type  = "cx11"
  ssh_username = "root"
}

build {
  source "hcloud.cfg" {
    # Set the snapshot "description", what is visible in the UI and from the
    # CLI with `hcloud image list | grep snapshot`.
    # Note that user-created images don't have the field "name".
    snapshot_name = "banana"
    # We need this kludge to refer to the image from Terraform, since the field
    # "name" is empty. This is not needed to show the bug of the Packer plugin,
    # but I keep it because without it it cannot be used by Terraform.
    snapshot_labels = {
      "name" : "banana"
    }
  }
}
```

### Operating system and Environment details

Doesn't apply.

### Log Fragments and crash.log files

Doesn't apply.

### PR DESCRIPTION
**Pull request by marco-m**
Before, it was possible to create multiple snapshots with the same name, leading to a confusing and wasteful situation.

Now, we perform the same kind of checks done by the AWS EC2 builder, and refuse to proceed if there is an existing snapshot with the same name.

Invoking `packer build -force` will bypass the checks, thus multiple snapshots with the same name will be allowed. This is equivalent to the previous behavior of this plugin.

Closes #14.

### CODE REVIEW

**Comment by marco-m** [comment]
Hello @nywilken, would you (or anybody from the Packer team) have time to look at this ? Thanks!

File: builder/hcloud/step_pre_validate.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+		ui.Say("Force flag found, skipping prevalidating snapshot name")
+		return multistep.ActionContinue
+	}
+
+	ui.Say(fmt.Sprintf("Prevalidating snapshot name: %s", s.SnapshotName))
+
+	// We would like to ask only for snapshots with a certain name using
+	// ImageListOpts{Name: s.SnapshotName}, but snapshots do not have name, they
+	// only have description. Thus we are obliged to ask for _all_ the snapshots.
+	opts := hcloud.ImageListOpts{Type: []hcloud.ImageType{hcloud.ImageTypeSnapshot}}
```
Comment (nywilken) [inline builder/hcloud/step_pre_validate.go:34]: My only concern here is the number of snapshots that could be obtained. Specifically, if there are too many snapshots would there be a need to paginate.  But it doesn't seem like there is any pagination for this method.

**Review by nywilken — APPROVED** [review]
Apologies for the delayed response. This looks great!

File: builder/hcloud/step_pre_validate.go (hunk 1 above)
Comment (marco-m) [reply builder/hcloud/step_pre_validate.go:34]: I agree, but I don't know what to do to address this.
Thinking about it, on the optimist side, if the list "overflows" and thus we get only the first page, the worst that can happen is that we don't detect that there is already a snapshot with the same "description", thus we will build it nonetheless, as the code was doing before. It is a sort of (involuntary ;-) fail safe.

**Pull request merged**